# check whether the stops in 2018 is the same with in 2020

In [2]:
import matplotlib.pyplot as plt
import matplotlib.mlab as mlab
from datetime import timedelta
from datetime import datetime
import pandas as pd
import numpy as np
import pymysql

In [3]:
#connect to the database
connection=pymysql.connect(host='taro.caxktie247nu.eu-west-1.rds.amazonaws.com'
                            ,port=3306
                            ,user='admin'
                            ,password='9Nw8v9yhx8uOb2nXVL8n'
                            ,db='dublinbus'
                            #,ssl={"require_secure_transport":'Off'}
                            ,cursorclass=pymysql.cursors.SSCursor)

#create a cur
cur = connection.cursor()

In [10]:
#get stops from the 2018 data
sql="select distinct `StopPointID` from dublinbus.Jan_LeaveTimes"
cur.execute(sql)
StopPointID=cur.fetchall()
stopPointID_18=[]
for stop in StopPointID:
    stopPointID_18.append(stop[0])
stopPointID_18

OperationalError: (2006, "MySQL server has gone away (ConnectionResetError(10054, '远程主机强迫关闭了一个现有的连接。', None, 10054, None))")

In [11]:
len(stopPointID_18)

4714

In [5]:
#load 2020 data
df_stops_20=pd.read_json("../data/stops.json")

In [6]:
stopPointID_20=df_stops_20.stops.values
stopPointID_20

array([   2,    3,    4, ..., 7709, 7710, 7711], dtype=int64)

In [12]:
len(stopPointID_20)

4145

In [7]:
def binary_search(item,alist):
    n = len(alist)
    start = 0
    end = n - 1
    while start <= end:
        mid = (start + end) // 2
        if alist[mid] == item:
            return False
        elif item < alist[mid]:
            end = mid - 1
        else:
            start = mid + 1
    return True

In [8]:
difference_18_20=[]
for stop in stopPointID_18:
    if binary_search(stop,stopPointID_20):
        difference_18_20.append(stop)
len(difference_18_20)

614

In [9]:
difference_18_20

[124,
 128,
 132,
 139,
 140,
 157,
 158,
 159,
 173,
 174,
 313,
 347,
 349,
 401,
 403,
 430,
 431,
 432,
 433,
 434,
 437,
 459,
 460,
 462,
 464,
 465,
 466,
 467,
 468,
 486,
 661,
 662,
 663,
 765,
 788,
 805,
 806,
 815,
 816,
 863,
 864,
 865,
 866,
 867,
 869,
 870,
 873,
 874,
 935,
 936,
 951,
 952,
 953,
 956,
 957,
 958,
 959,
 960,
 961,
 963,
 964,
 965,
 966,
 967,
 968,
 969,
 973,
 974,
 975,
 976,
 979,
 988,
 989,
 990,
 991,
 997,
 998,
 999,
 1000,
 1001,
 1002,
 1003,
 1004,
 1006,
 1007,
 1008,
 1009,
 1010,
 1012,
 1039,
 1040,
 1041,
 1050,
 1131,
 1132,
 1133,
 1134,
 1138,
 1139,
 1252,
 1253,
 1391,
 1392,
 1394,
 1395,
 1400,
 1421,
 1513,
 1530,
 1546,
 1591,
 1601,
 1602,
 1603,
 1645,
 1650,
 1651,
 1652,
 1766,
 1820,
 1824,
 1825,
 1830,
 1831,
 1856,
 1857,
 1880,
 1881,
 2027,
 2028,
 2029,
 2030,
 2037,
 2038,
 2047,
 2048,
 2049,
 2050,
 2052,
 2071,
 2072,
 2073,
 2074,
 2075,
 2076,
 2079,
 2080,
 2082,
 2083,
 2110,
 2111,
 2112,
 2113,
 2116,


there 614 stops that exists in 2018 but not in 2020

In [62]:
difference_20_18=[]
for stop in stopPointID_20:
    if binary_search(stop,stopPointID_18):
        difference_20_18.append(stop)
len(difference_20_18)

45

there 45 stops that exists in 2020 but not in 2018

if stops in 2018 but not in 2020, it means when training the model, it should ignore the time during in this stop(dep-arr).  
if stops in 2020 but not in 2018, discuss.